File: Hierarchical.py

class Human:
    name = ""

    def __init__(self, name):
        self.name = name

    def func(self):
        print(f"{self.name} inside the func")

class Male(Human):
    gender = ""

    def __init__(self, name, gender):
        self.gender = gender
        super().__init__(name)

    def male_func(self):
        print(f"{self.name} gender is : {self.gender}")

class Female(Human):
    gender = ""

    def __init__(self, name, gender):
        self.gender = gender
        super().__init__(name)

    def female_func(self):
        print(f"{self.name} gender is : {self.gender}")


print("human class")
human = Human("super human")
print(human.name)
human.func()


print("male class")
print()
male = Male("sumon", "male")
print(male.name)
male.func()
male.male_func()


print("female class")
print()
female = Female("s***", "female")
print(female.name)
female.func()
female.female_func()

File: Hierarchical.cpp
#include<bits/stdc++.h>
using namespace std;

class Human{
    public:
        string name;

        // Human(string name){
        //     this->name = name;
        // }

        void name_func(){
            cout<<"Name is : "<<name<<endl;
        }
};
class Male: public Human{
    public:
        string gender;

        // Male(string gender){
        //     this->gender = gender;
        // }

        void male_gender(){
            cout<<name<<" Gender is "<<gender<<endl;
        }
};
class Female: public Human{
    public:
        string gender;

        // Female(string gender){
        //     this->gender = gender;
        // }

        void female_gender(){
            cout<<name<<" Gender is "<<gender<<endl;
        }
};

int main(){

    Human h;
    h.name = "super human";
    cout<<h.name<<endl;
    h.name_func();

    Male s;
    s.gender = "male";
    s.name = "Sumon";

    Female S;
    S.gender = "female";
    S.name = "S**********";

    cout<<s.name<<endl;
    s.male_gender();

    cout<<S.name<<endl;
    S.female_gender();

    return 0;
}